genMatrix.js

```diff
@@ -0,0 +1,82 @@
+'use strict';
+const path = require('path');
+const fs = require('fs');
+
+const testFiles = [
+  'genMatrix.js',
+  '.github/workflows/build-test.yml',
+];
+
+const nodeDirRegex = /^\d+$/;
+
+const areTestFilesChanged = (changedFiles) => changedFiles
+  .filter((file) => testFiles.includes(file)).length;
+
+// Returns a list of the child directories in the given path
+const getChildDirectories = (parent) => fs.readdirSync(parent, { withFileTypes: true })
+  .filter((dirent) => dirent.isDirectory())
+  .map(({ name }) => path.resolve(parent, name));
+
+const getNodeVerionDirs = (base) => getChildDirectories(base)
+  .filter((childPath) => nodeDirRegex.test(path.basename(childPath)));
+
+// Returns the paths of Dockerfiles that are at: base/*/Dockerfile
+const getDockerfilesInChildDirs = (base) => getChildDirectories(base)
+  .map((childDir) => path.resolve(childDir, 'Dockerfile'));
+
+const getAllDockerfiles = (base) => getNodeVerionDirs(base).flatMap(getDockerfilesInChildDirs);
+
+const getAffectedDockerfiles = (filesAdded, filesModified, filesRenamed) => {
+  const files = [
+    ...filesAdded,
+    ...filesModified,
+    ...filesRenamed,
+  ];
+
+  // If the test files were changed, include everything
+  if (areTestFilesChanged(files)) {
+    console.log('Test files changed so scheduling all Dockerfiles');
+    return getAllDockerfiles(__dirname);
+  }
+
+  const modifiedDockerfiles = files.filter((file) => file.endsWith('/Dockerfile'));
+
+  // Get Dockerfiles affected by modified docker-entrypoint.sh files
+  const entrypointAffectedDockerfiles = files
+    .filter((file) => file.endsWith('/docker-entrypoint.sh'))
+    .map((file) => path.resolve(path.dirname(file), 'Dockerfile'));
+
+  return [
+    ...modifiedDockerfiles,
+    ...entrypointAffectedDockerfiles,
+  ];
+};
+
+const getFullNodeVersionFromDockerfile = (file) => fs.readFileSync(file, 'utf8')
+  .split('\n')
+  .find((line) => line.startsWith('ENV NODE_VERSION '))
```

Comment:
Think you might be able single regex instead of spliting. EX: something like `/^ENV NODE_VERSION (\d*\.*\d*\.\d*)/`

Reply:
Was this what you had in mind? https://github.com/nodejs/docker-node/pull/1339/files#diff-779f7d1132d6c93dd0b52d5bf161b162R55

```js
const getFullNodeVersionFromDockerfile = (file) => fs.readFileSync(file, 'utf8')
  .match(/^ENV NODE_VERSION (\d*\.*\d*\.\d*)/m)[1];
```

Reply:
If that's a working version of the regex, then yes 😆 
Probably a micro-optimization at best